transforms/universal/fdedup/python/src/cluster_analysis_transform.py

```diff
@@ -0,0 +1,321 @@
+        num_segments: the number of segments dividing the hashing space for each band
+    """
+
+    def __init__(self, config: dict[str, Any]):
+        """
+        Initialize based on the dictionary of configuration information.
+        This is generally called with configuration parsed from the CLI arguments
+        defined by the companion runtime, ClusterAnalysisTransformRuntime.
+        """
+        super().__init__(config)
+        self.num_bands = config.get(num_bands_key, num_bands_default)
+        self.num_segments = config.get(num_segments_key, num_segments_default)
+        self.jaccard_similarity_threshold = config.get(
+            jaccard_similarity_threshold_key, jaccard_similarity_threshold_default
+        )
+        self.data_access = config.get("data_access")
+        self.logger = get_logger(__name__)
+
+    def transform(self, folder_name: str) -> tuple[list[tuple[bytes, str]], dict[str, Any]]:
+        self.logger.info(f"Cluster analysis for folder {folder_name}")
+        metadata = {}
+        input_folder = self.sanitize_folder_name(os.path.join(self.data_access.input_folder, folder_name))
+        files, retries = self.data_access.get_folder_files(
+            path=input_folder,
+            extensions=[".parquet"],
+            return_data=True,
+        )
+        if retries > 0:
+            metadata |= {"data_access_retries": retries}
+        match = re.match(r"^band=(\d+)/segment=(\d+)$", folder_name)
+        if match:
+            band = int(match.group(1))
+            segment = int(match.group(2))
+        else:
+            raise ValueError(f"Wrong folder_name {folder_name}, should be band=b/segment=s")
+        output_folder = self.sanitize_folder_name(self.data_access.output_folder)
+        output_path = os.path.join(output_folder, f"band_{band}_segment_{segment}.parquet")
+
+        # consolidate into a single data frame band hashes computed by workers
+        band_segment_dataframe, consolidation_stats = self.consolidate_band_segment_files(files)
+        metadata |= consolidation_stats
+        # cluster grouping by band hashes
+        cluster_dataframe, cluster_stats = self.get_clusters(band_segment_dataframe)
+        metadata |= cluster_stats
+        # cluster analysis using jaccard similarity
+        jaccard_cluster_dataframe, jaccard_stats = self.analyze_clusters(cluster_dataframe)
+        metadata |= jaccard_stats
+        # Generate the docs_to_remove dataframe
+        docs_to_remove_dataframe = jaccard_cluster_dataframe.explode("docs_to_remove")
+        output_data = TransformUtils.convert_arrow_to_binary(docs_to_remove_dataframe.to_arrow())
+        self.logger.info(f"{len(docs_to_remove_dataframe)} documents marked to remove")
+        metadata |= {"num_duplicate_documents": len(docs_to_remove_dataframe)}
+        return [(output_data, output_path)], metadata
+
+    def sanitize_folder_name(self, folder_name: str) -> str:
```

Comment:
This looks like the same method as in data_cleaning_transform.py.  Is it worth having a single static or non-class utility method somewhere?